chain/chain/src/stateless_validation/chunk_validation.rs

```diff
@@ -113,74 +107,204 @@ pub fn validate_prepared_transactions(
     )
 }
 
+pub enum ImplicitTransitionParams {
+    ApplyChunk(ApplyChunkBlockContext, ShardContext),
+    Resharding(AccountId, RetainMode, ShardUId),
+}
+
 struct StateWitnessBlockRange {
-    /// Blocks from the last new chunk (exclusive) to the parent block
-    /// (inclusive).
-    blocks_after_last_chunk: Vec<Block>,
+    /// Transition parameters **after** the last chunk, corresponding to all
+    /// state transitions from the last new chunk (exclusive) to the parent
+    /// block (inclusive).
+    implicit_transition_params: Vec<ImplicitTransitionParams>,
     /// Blocks from the last last new chunk (exclusive) to the last new chunk
-    /// (inclusive).
+    /// (inclusive). Note they are in **reverse** order, from the newest to the
+    /// oldest. They are needed to validate the chunk's source receipt proofs.
     blocks_after_last_last_chunk: Vec<Block>,
+    /// Shard id and index of last chunk before the chunk being validated.
     last_chunk_shard_id: ShardId,
     last_chunk_shard_index: ShardIndex,
 }
 
+/// Gets ranges of blocks that are needed to validate a chunk state witness.
+/// Iterates backwards through the chain, from the chunk being validated to
+/// the second last chunk, if it exists.
 fn get_state_witness_block_range(
     store: &ChainStore,
     epoch_manager: &dyn EpochManagerAdapter,
     state_witness: &ChunkStateWitness,
 ) -> Result<StateWitnessBlockRange, Error> {
-    let mut blocks_after_last_chunk = Vec::new();
+    let mut implicit_transition_params = Vec::new();
     let mut blocks_after_last_last_chunk = Vec::new();
 
-    let mut block_hash = *state_witness.chunk_header.prev_block_hash();
-    let mut prev_chunks_seen = 0;
+    /// Position in the chain while traversing the blocks backwards.
+    struct TraversalPosition {
+        /// Shard ID of chunk, needed to validate state transitions, in the
+        /// currently observed block.
+        shard_id: ShardId,
+        /// Hash of the previous block.
+        prev_hash: CryptoHash,
+        /// Previous block.
+        prev_block: Block,
+        /// Number of chunks seen during traversal.
+        chunks_seen: u32,
+        /// Below, supposed id and index of last chunk before the chunk being
+        /// validated.
```

Comment:
nit: I can't quite parse this sentence, can you put in different words? The "below" in particular is confusing.